src/TSTransformer/nodes/class/transformDecorators.ts

```diff
@@ -4,28 +4,81 @@ import { TransformState } from "TSTransformer";
 import { transformExpression } from "TSTransformer/nodes/expressions/transformExpression";
 import { transformPropertyName } from "TSTransformer/nodes/transformPropertyName";
 import { convertToIndexableExpression } from "TSTransformer/util/convertToIndexableExpression";
+import { expressionMightMutate } from "TSTransformer/util/expressionMightMutate";
 import ts from "typescript";
 
+type HasDecorators = Exclude<ts.HasDecorators, ts.AccessorDeclaration>;
+
+function countDecorators(node: HasDecorators) {
+	return ts.getDecorators(node)?.length ?? 0;
+}
+
+function findConstructor(node: ts.ClassLikeDeclaration): ts.ConstructorDeclaration | undefined {
+	return node.members.find((member): member is ts.ConstructorDeclaration => {
+		return ts.isConstructorDeclaration(member) && member.body !== undefined;
+	});
+}
+
+function shouldInline(
+	state: TransformState,
+	isLastDecorator: boolean,
+	decorator: ts.Decorator,
+	expression: luau.Expression,
+): boolean {
+	// immutable expressions can be inlined
+	if (!expressionMightMutate(state, expression, decorator.expression)) return true;
+
+	// if it's not the last decorator, we can't inline
+	// this is because we need to initialize all decorators before running them
+	if (!isLastDecorator) return false;
+
+	const node = decorator.parent;
+
+	// if the node is a method declaration and has a decorator on a parameter, we can't inline
```

Comment:
These comments are basically the same as the code, and I don't think the code is difficult to read here. Comments should explain why, not (just) what.